lms/djangoapps/instructor_task/tasks_helper/certs.py

```diff
@@ -142,12 +153,20 @@ def _invalidate_generated_certificates(course_id, enrolled_students, certificate
         status__in=certificate_statuses,
     )
 
+    allowlisted_users = get_allowlisted_users(course_id)
+
     # Mark generated certificates as 'unavailable' and update download_url, download_uui, verify_uuid and
-    # grade with empty string for each row
-    certificates.update(
-        status=CertificateStatuses.unavailable,
-        verify_uuid='',
-        download_uuid='',
-        download_url='',
-        grade='',
-    )
+    # grade with empty string for each cert that is not allowlisted. We loop over the certs and save each individually
+    # in order to save a history of the change.
+    for c in certificates:
+        if c.user in allowlisted_users:
+            log.info(f'Certificate for user {c.user.id} will not be invalidated because they are on the allowlist for '
+                     f'course {course_id}')
+        else:
+            log.info(f'About to invalidate certificate for user {c.user.id} in course {course_id}')
```

Comment:
Would we want to call the [invalidate](https://github.com/edx/edx-platform/blob/42fc6aef035334524b8caaba064ac3cfb6756fe3/lms/djangoapps/certificates/models.py#L365) method of the GeneratedCertificate model here for each certificate in this loop?

This would ensure we emit the `COURSE_CERT_REVOKED` signal as well which I believe will help fix some data inconsistency issues on the Credentials IDA side.

Reply:
I was planning on doing that in a subsequent PR. 

Because this code invalidates a cert and then immediately regenerates it, I wasn't sure if we could rely on those 2 signals to not step on each other. I thought being able to monitor it separately from any allowlist changes could be useful.

Reply:
That makes sense to me. I hadn't thought about the potential for conflicting signals!